### PR TITLE
ARROW-10320 [Rust] [DataFusion] Migrated from batch iterators to batch streams.

### PR DESCRIPTION
**Pull request by jorgecarleitao**
Recently, we introduced `async` to `execute`. This allowed us to parallelize multiple partitions as we denote an execution of a part (of a partition) as the unit of work. However, a part is often a large task composing multiple batches and steps.

This PR makes all our execution nodes return a dynamically-typed [`Stream<Item = ArrowResult<RecordBatch>>`](https://docs.rs/futures/0.3.6/futures/stream/trait.Stream.html) instead of an Iterator. For reference, a Stream is an iterator of futures, which in this case is a future of a `Result<RecordBatch>`.

This effectively breaks the execution in smaller units of work (on which an individual unit is an operation returns a `Result<RecordBatch>`) allowing each task to chew smaller bits.

This adds `futures` as a direct dependency of DataFusion (it was only a dev-dependency).

This leads to a +2% degradation in aggregates in micro benchmarking, which IMO is expected given that there is more context switching to handle. However, I expect (hope?) this to be independent of the number of batches and partitions, and be offset by any async work we perform to our sources (readers) and sinks (writers).

I did not take the time to optimize - the primary goal was to implement the idea, have it compile and pass tests, and have some discussion about it. I expect that we should be able to replace some of our operations by `join_all`, thereby scheduling multiple tasks at once (instead of waiting one by one).

<details>
 <summary>Benchmarks</summary>

Aggregates:
```
aggregate_query_no_group_by 15 12                                                                            
                        time:   [782.11 us 784.12 us 786.19 us]
                        change: [+1.1421% +2.5252% +3.8963%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe

aggregate_query_group_by 15 12                                                                            
                        time:   [5.8751 ms 5.9206 ms 5.9679 ms]
                        change: [+1.0645% +2.0027% +3.0333%] (p = 0.00 < 0.05)
                        Performance has regressed.

aggregate_query_group_by_with_filter 15 12                                                                             
                        time:   [2.7652 ms 2.7983 ms 2.8340 ms]
                        change: [+0.3278% +1.8981% +3.3819%] (p = 0.02 < 0.05)
                        Change within noise threshold.
```

Math:
```
sqrt_20_9               time:   [6.9844 ms 7.0582 ms 7.1363 ms]                      
                        change: [+0.0557% +1.5625% +3.0408%] (p = 0.05 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

sqrt_20_12              time:   [2.8350 ms 2.9504 ms 3.1204 ms]                        
                        change: [+3.8751% +8.2857% +14.671%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe

sqrt_22_12              time:   [14.888 ms 15.242 ms 15.620 ms]                       
                        change: [+7.6388% +10.709% +14.098%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

sqrt_22_14              time:   [23.710 ms 23.817 ms 23.953 ms]                       
                        change: [-4.3401% -3.1824% -2.0952%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  5 (5.00%) high mild
  6 (6.00%) high severe
```
</details>

I admit that this is a bit outside my comfort zone, and someone with more experience in `async/await` could be of help.

IMO this would integrate very nicely with ARROW-10307, ARROW-9275, I _think_ it would also help ARROW-9707, and I _think_ that it also opens the possibility consuming / producing batches from/to sources and sinks from arrow-flight / IPC

### CODE REVIEW

**Comment by jorgecarleitao** [comment]
FYI @alamb @vertexclique @maxburke

**Comment by andygrove** [comment]
This is great to see. I ran a quick TPC-H query 1 benchmark against the 100 GB data set with this PR and saw no noticeable difference in performance, so the microbenchmark might not be so relevant here.

File: rust/datafusion/src/physical_plan/hash_aggregate.rs
Comment (alamb) [inline rust/datafusion/src/physical_plan/hash_aggregate.rs:571]: this is cool -- to incrementally accumulate the aggregates in the strems

File: rust/datafusion/src/physical_plan/merge.rs
Comment (alamb) [inline rust/datafusion/src/physical_plan/merge.rs:110]: I am still concerned that these calls to `collect` effectively buffer all the input before producing any output -- this both will likely consume more memory than needed as well as being a 'pipeline breaker' (nothing that relies on the output of the `Merge` can run until *all* of the merge inputs have been produced). 

We could peraps use `chain` here to fuse the streams together -- https://docs.rs/tokio/0.3.0/tokio/stream/trait.StreamExt.html#method.chain. Using chain would avoid the need to buffer the intermediate results (aka the collect), but  it would also likely cause the input partitions to run one after the other rather than in parallel_

Another thought I had would be to use something fancier like a `channel` that all the substreams write to. 

But in any event that could be done in some future PR -- I think this code is better than master, and a step forward.

File: rust/datafusion/src/physical_plan/parquet.rs
Comment (alamb) [inline rust/datafusion/src/physical_plan/parquet.rs:212]: This is getting closer to `async` parquet reader 👍

File: rust/datafusion/tests/user_defined_plan.rs
Comment (alamb) [inline rust/datafusion/tests/user_defined_plan.rs:528]: this is very cool to see how the user defined code got transformed to a `Stream`

**Review by alamb — APPROVED** [review]
FWIW I really like this approach. Thanks @jorgecarleitao

File: rust/datafusion/src/datasource/memory.rs
Comment (vertexclique) [reply rust/datafusion/src/datasource/memory.rs:137]: Please not do that. Futures interfaces are not compatible with other APIs in the ecosystem. This is a good implementation which is adaptable to tokio, async-std, bastion and others.

File: rust/datafusion/src/physical_plan/hash_aggregate.rs
Comment (vertexclique) [inline rust/datafusion/src/physical_plan/hash_aggregate.rs:398]: Can you use `map` instead of pattern matching? since err escalation looks redundant.

File: rust/datafusion/src/physical_plan/hash_aggregate.rs
Comment (vertexclique) [inline rust/datafusion/src/physical_plan/hash_aggregate.rs:591]: Same map simplification here probably.

File: rust/datafusion/src/physical_plan/hash_aggregate.rs
Comment (vertexclique) [inline rust/datafusion/src/physical_plan/hash_aggregate.rs:585]: 2 is `to` i presume. Right :) ?

**Review by vertexclique — COMMENTED** [review]
Small nits. Apart from that, it is looking good!

**Comment by jorgecarleitao** [comment]
I am still not entirely happy with this, and I think that we need a further step to make this work well, which is the reason I placed this back to draft.

Essentially, I misinterpreted what a `Stream` is used for. A `stream` of record batches will only return the next batch _after_ the previous one was (async) computed. Therefore, we can't parallelize, as we are blocked by the execution of the previous batch. Therefore, this PR is only useful for situations on which we are reading batches that, we may need to wait for.

I think that the solution is to put a `future` as the `Item` of the `Stream`, so that we can collect all those futures and perform `join_all().await`, thereby sending every task to the scheduler.

PR #8480 is the proposal for that. I doubles the performance of the larger projections benchmarks (purely due to scheduling / threading).

Note that each PR solves a different problem: this one is related to the fact that some batches may not be available yet (and thus we should not block), the other one is related to the fact that some operations can be parallelized (and thus we should `spawn`). I think that together they address the main issues.

**Comment by andygrove** [comment]
> A `stream` of record batches will only return the next batch _after_ the previous one was (async) computed. Therefore, we can't parallelize, as we are blocked by the execution of the previous batch.

That sounds like exactly what we need. We compute partitions in parallel and each partition produces an ordered set of batches. In some cases, it is critical that the order is deterministic.

File: rust/datafusion/src/physical_plan/merge.rs
Comment (andygrove) [reply rust/datafusion/src/physical_plan/merge.rs:110]: Yes, collect is the reason we don't see better performance from this PR IMO. I agree that this can be fixed in a separate PR.

**Comment by alamb** [comment]
> A stream of record batches will only return the next batch after the previous one was (async) computed. Therefore, we can't parallelize, as we are blocked by the execution of the previous batch. Therefore, this PR is only useful for situations on which we are reading batches that we may need to wait for.

@jorgecarleitao  -- I personally think leaving the interface to be `Stream<RecordBatch>` is the way to go. 

Among other things, this interface allows "backpressure" so that we don't end up buffering intermediate results if one operator (e.g. filter) can produce record batches than its consumer can process them (e.g. aggregate). 

If we want to add additional parallelism with a tunable cost of additional buffering, we could keep the interface of this PR and add an operator like "Buffer". Buffer would calculate and store some number of `RecordBatches`, perhaps in a channel, as separate tasks in advance of when a batch was requested. 

Then the user of DataFusion could control how much more buffering they were willing to accept in order to get more parallelism. 

I think an approach like https://github.com/apache/arrow/pull/8480 is cool, but the choice of the aggregate, for example, to compute the input as fast as possible in new tasks, even if the aggregate

File: rust/datafusion/src/physical_plan/hash_aggregate.rs
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_aggregate.rs:585]: it is `2.`, referring to the item number two in the list of items above.

**Comment by jorgecarleitao** [comment]
I agree with you that #8480 has other issues that we want to avoid. Therefore, I placed this back as ready to review.

Thanks a lot for the comments so far: this concurrency stuff is new to me and thus your guidance is really appreciated here.

File: rust/datafusion/src/physical_plan/common.rs
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/common.rs:43]: I was thinking to reserve `Buffered` to a stream that buffers N entries. This one is already initialized with everything via `new`. This is more like the [`iter`](https://docs.rs/futures/0.3.6/futures/stream/fn.iter.html), with a fixed known len and a known schema. IMO we should avoid using this struct, as it requires us to collect to it somewhere.

Comment (alamb) [reply rust/datafusion/src/physical_plan/common.rs:43]: Yeah, I was thinking it would be nice to imply in this struct's name somehow that it should be avoided if possible, because, as you say, it requires buffering / collecting up the `RecordBatch`es before producing any results

**Review by andygrove — APPROVED** [review]
LGTM

File: rust/datafusion/src/physical_plan/hash_aggregate.rs
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_aggregate.rs]: Good point. Fixed (and below).

File: rust/datafusion/src/physical_plan/merge.rs
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/merge.rs:110]: PR here: https://github.com/apache/arrow/pull/8503

**Pull request closed without merging**
